### DISCUSSION TITLE
How to update a deeply nested array?

### DISCUSSION THREAD

**Discussion by grantjayy** [prisma/prisma#14586] [Q&A]
Hey there,

I'm trying to update an array in a deeply nested object with the Prisma client using MongoDB.

For my life, I cannot figure out how to PUSH to the array to add one element to it.

I have read the scalar list documentation, the push API reference, and still nothing.

Here's a sample of my scheme:

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

type TrackingUtm {
  mediums String[]
}

type Tracking {
  utm TrackingUtm?
}

model users {
  id           String    @id @default(auto()) @map("_id") @db.ObjectId
  created_at   DateTime  @default(now())
  tracking     Tracking?
}
```

This is the code that I think should work:

```ts
user = await prisma.users.update({
			where: { id: user.id },
			data: {
				tracking: {
					utm: {
						mediums: {
							push: 'a'
						}
					}
				}
			}
		});
```

Here's the error I see:
<img width="936" alt="Screen Shot 2022-07-31 at 8 12 26 PM" src="https://user-images.githubusercontent.com/47218341/182059165-0be5e360-bbf6-4b14-86a0-b15beb7415fa.png">

Even if I try to set the value it throws an error:

```ts
user = await prisma.users.update({
			where: { id: user.id },
			data: {
				tracking: {
					utm: {
						mediums: {
							set: 'a'
						}
					}
				}
			}
		});
```

```
Unknown arg `utm` in data.tracking.utm for type TrackingNullableUpdateEnvelopeInput. Did you mean `set`? Available args:
type TrackingNullableUpdateEnvelopeInput {
set?: TrackingCreateInput | Null
upsert?: TrackingUpsertInput
unset?: Boolean
```

Can somebody please explain to me what I am doing wrong?

**Comment by austincrim** [comment]
Hey @grantjayy 👋,

The underlying cause here is because your schema includes **optional composite type**. Since the field is optional, Prisma needs to know how to handle the scenario in which the optional fields don't exist.

The way around this is to use `upsert` for any optional fields.
```ts
const user = await prisma.users.update({
  where: { id: '1' },
  data: {
    tracking: {
      upsert: {
        set: {
            utm: {}
        },
        update: {
          utm: {
            upsert: {
              set: {
                  mediums: {
                      set: [],
                  }
              },
              update: {
                  mediums: {
                      push: ['a']
                  },
                }
            }
          }
        }
      }
    }
  }
})
```

We note this [here in the docs](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#remarks-30).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
